File: v25_iterator/java/StudentMain.java
package v25_iterator.java;

import java.util.ArrayList;
import java.util.List;

interface Iterator {
    boolean hasNext();
    Object next();
}

class Student {
    private String name;
    private int age;

    public Student(String name, int age) {
        this.name = name;
        this.age = age;
    }

    public String getName() {
        return name;
    }

    public int getAge() {
        return age;
    }
}

class ClassroomIterator implements Iterator {
    private Classroom classroom;
    private int index;

    public ClassroomIterator(Classroom classroom) {
        this.classroom = classroom;
        this.index = 0;
    }

    @Override
    public boolean hasNext() {
        return index < classroom.getSize();
    }

    @Override
    public Object next() {
        Student student = classroom.getStudentAt(index);
        index++;
        return student;
    }
}


interface Aggregate {
    Iterator createIterator();
}

class Classroom implements Aggregate {
    private List<Student> students;

    public Classroom() {
        students = new ArrayList<>();
    }

    public void addStudent(Student student) {
        students.add(student);
    }

    public Student getStudentAt(int index) {
        return students.get(index);
    }

    public int getSize() {
        return students.size();
    }

    @Override
    public Iterator createIterator() {
        return new ClassroomIterator(this);
    }
}



public class StudentMain {

    public static void main(String[] args) {
        Classroom classroom = new Classroom();
        classroom.addStudent(new Student("Alice", 20));
        classroom.addStudent(new Student("Bob", 21));
        classroom.addStudent(new Student("Charlie", 22));

        Iterator iterator = classroom.createIterator();
        while (iterator.hasNext()) {
            Student student = (Student) iterator.next();
            System.out.println(student.getName() + ": " + student.getAge());
        }
    }
}

File: v25_iterator/python/student_main.py
from collections.abc import Iterator

class Student:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def get_name(self):
        return self.name

    def get_age(self):
        return self.age

class ClassroomIterator(Iterator):
    def __init__(self, classroom):
        self.classroom = classroom
        self.index = 0

    def __next__(self):
        try:
            student = self.classroom.get_student_at(self.index)
        except IndexError:
            raise StopIteration()
        self.index += 1
        return student

class Classroom:
    def __init__(self):
        self.students = []

    def add_student(self, student):
        self.students.append(student)

    def get_student_at(self, index):
        return self.students[index]

    def get_size(self):
        return len(self.students)

    def create_iterator(self):
        return ClassroomIterator(self)

def main():
    classroom = Classroom()
    classroom.add_student(Student("Alice", 20))
    classroom.add_student(Student("Bob", 21))
    classroom.add_student(Student("Charlie", 22))

    iterator = classroom.create_iterator()
    while True:
        try:
            student = next(iterator)
            print(student.get_name() + ": " + str(student.get_age()))
        except StopIteration:
            break

if __name__ == "__main__":
    main()


